lib/api/core/entrypoints/embedded/protocols/websocket.js

```diff
@@ -217,34 +246,68 @@ class WebsocketProtocol extends Protocol {
         }
       };
 
-      this.entryPoint.kuzzle.pluginsManager.trigger('log:error', new KuzzleInternalError(e.message));
+      this.entryPoint.kuzzle.pluginsManager.trigger(
+        'log:error', new KuzzleInternalError(e.message));
 
       return this._send(connection.id, JSON.stringify(errobj));
     }
   }
 
+  /**
+   * /!\ WARNING: CRITICAL CODE SECTION AHEAD /!\
+   * (this means performance over maintenability or readability... sorry)
+   *
+   * Do not change without reason: this method is used by the real-time engine
+   * to send very similar notifications to A LOT of different sockets
+   *
+   * This function precomputes RFC-6455 WebSocket frames and send them raw to
+   * destination sockets. This prevents the recomputing of the same frame
+   * over and over again for each broadcast.
+   *
+   * @param  {Object} data
+   */
   broadcast(data) {
-    /*
-     Avoids stringifying the payload multiple times just to update the room:
-     - we start deleting the last character, which is the closing JSON bracket ('}')
-     - we then only have to inject the following string to each channel:
-     ,"room":"<roomID>"}
+    const
+      stringified = JSON.stringify(data.payload),
+      // 255 bytes should be enough to hold the following:
+      //     ,"room":"<channel identifier>"
+      // (with current channel encoding, this is less than 100 bytes)
+      payload = Buffer.allocUnsafe(stringified.length + 255);
+
+    let offset = stringified.length - 1;
+
+    payload.write(stringified, 0);
+    payload.write(jsonRoomProp, offset);
+
+    offset += jsonRoomProp.length;
 
-     So, instead of stringifying the payload for each channel, we only concat
-     a new substring to the original payload.
-     */
-    const payload = JSON.stringify(data.payload).slice(0, -1) + ',"room":"';
+    for (const channel of data.channels) {
+      const connectionIds = this.channels.get(channel);
 
-    debug('broadcast: %a', data);
+      if (connectionIds) {
+        // Adds the room property to the message
+        payload.write(channel, offset);
+        payload.write(jsonEnder, offset + channel.length);
 
-    for (const channelId of data.channels) {
-      if (this.channels[channelId]) {
+        // prevent buffer overwrites due to socket.send being an
```

Comment:
This is explained in the comment, but I'll detail a bit more here:
* the `payload` buffer is reused over and over, to reduce the amount of memory manipulations
* socket.send being an asynchronous method by nature (I/O are made on separate threads)

This means that if I use `payload` directly (by creating a DataView on top of it obviously), then by the time the frame is actually sent, the buffer might contain another message.

As stated, we need to perform a full copy to prevent that race condition to occur.

Also, copying a buffer content to another buffer is akin to a low-level `memcpy` operation

Reply:
Yes but you send directly the frame with `_sender.sendFrame` and not the payload directly.
As far I can see, the data is already copied: https://github.com/websockets/ws/blob/master/lib/sender.js#L60

Reply:
Nope: `frame` merges the data frame only if we use a mask, which we do not: this is a bit costly and that only makes sense for clients sending data to a server to prevent proxy attacks.